spec_util/meld_test.go

```diff
@@ -271,6 +272,26 @@ var tests = []testData{
 		MeldOptions{},
 		"testdata/meld/meld_map_4_map_5_expected.pb.txt",
 	},
+	{
+		// Test meld(T, nullable<T>) => nullable<T>
+		"meld nullable and non-nullable versions of the same type",
+		[]string{
+			"testdata/meld/meld_nullable_1.pb.txt",
+			"testdata/meld/meld_nullable_2.pb.txt",
+		},
+		MeldOptions{},
+		"testdata/meld/meld_nullable_expected.pb.txt",
+	},
+	{
+		// Test meld(None, optional<T>) => nullable<optional<T>>
```

Comment:
Should this be an `optional<nullable<T>>` instead?

Reply:
Thinking on this more, in principle, `nullable<optional<T>>` doesn't make sense as a type: a `nullable<T>` isn't optional; it must be always present, even if null. Given this, I suppose we could let `nullable<optional<T>>` mean the same thing as `optional<nullable<T>>`, but having two representations of the same thing is a recipe for future headache.

This is just more evidence of a bad IR design. I don't know that there's much we can do here, beyond perhaps settling on a canonical representation of an optional nullable type.

Reply:
Yea, it's not ideal.  The situation is worse for `oneof<T, ...>`, because there are combinatorially many equivalent types: `nullable<oneof<T, ...>>` and any combination of nullable children.  I chose to attach the nullable bit to the topmost IR node, which is why `nullable<optional<T>>`.  (Although I just realized that I need to lift the nullable bit when combining/creating oneofs.)

If I were to redesign things, I would make required/optional a property of structs rather than a first class type.  After all, what does `oneof<optional<T>, T2>` mean, or `list<optional<T>>`?  We could get rid of `optional` entirely and add a `null` type inhabited by the `null` value.  Then `optional<T>` would be syntactic sugar for `oneof<T, null>`, and we could flatten `oneof`s to get canonical types.

Reply:
641cf703e351c4e5e6f72394f08f57c7b87ca55b updates the comment on `meldData` to explain the nullable invariant, and 260b2c8476adb4c31572fc8b7a156a9e1589f5eb ensures the invariant is maintained when melding optional and oneof nodes.

The latter also adds special handling for `Optional<None>`, which represents an optional field that, when present, has only been seen with a null value.